app/gui2/shared/languageServerTypes.ts

```diff
@@ -255,28 +250,65 @@ type FileSystemObject =
 
 interface VisualizationContext {}
 
+export interface VisualizationConfiguration {
+  /** An execution context of the visualization. */
+  executionContextId: Uuid
+  /** A qualified name of the module to be used to evaluate the arguments for the visualization
+   * expression. */
+  visualizationModule: string
+  /** An expression that creates a visualization. */
+  expression: string | MethodPointer
+  /** A list of arguments to pass to the visualization expression. */
+  positionalArgumentsExpressions?: string[]
+}
+
+export interface VCSSave {
+  commitId: string
+  message: string
+}
+
 export type Notifications = {
-  'file/event': [{ path: Path; kind: FileEventKind }]
-  'text/autoSave': [{ path: Path }]
-  'text/didChange': [{ edits: FileEdit[] }]
-  'text/fileModifiedOnDisk': [{ path: Path }]
-  'executionContext/expressionUpdates': [{ contextId: ContextId; updates: ExpressionUpdate[] }]
-  'executionContext/executionFailed': [{ contextId: ContextId; message: string }]
-  'executionContext/executionComplete': [{ contextId: ContextId }]
-  'executionContext/executionStatus': [{ contextId: ContextId; diagnostics: Diagnostic[] }]
-  'search/suggestionsDatabaseUpdate': [{}]
-  'file/rootAdded': [{}]
-  'file/rootRemoved': [{}]
-  'executionContext/visualizationEvaluationFailed': [
-    {
-      contextId: ContextId
-      visualizationId: Uuid
-      expressionId: ExpressionId
-      message: String
-      diagnostic?: Diagnostic
-    },
-  ]
-  'refactoring/projectRenamed': [{}]
+  'text/autoSave': (param: { path: Path }) => void
```

Comment:
I believe this can be spelled out with a bit shorter syntax:

```suggestion
  'text/autoSave'(param: { path: Path }): void
```

Reply:
yes, but method syntax is unsound: [ts handbook](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html#note)
(it's an old doc, but it's still true today)

Reply:
Wow. I didn't know about that. TIL. Why did they keep it bivariant in that case? We don't have any generics in argument position in this case, but we may as well do it the "less broken" way.

Reply:
apparently it's something like, so many people already use inheritance incorrectly that they figured it's not really reasonable to change it at that point.

theoretically they could add a new compiler option, but apparently they tend to avoid it if the feature isn't deemed very useful - i think it's understandable that adding even more complexity to the config wouldn't be the first thing they reach for